infrastructure/src/main/java/org/corfudb/infrastructure/log/LogMetadata.java

```diff
@@ -68,6 +85,19 @@ public void update(LogData entry) {
                 // but that can be addressed in another issue.
                 long currentStreamTail = streamTails.getOrDefault(streamId, Address.NON_ADDRESS);
                 streamTails.put(streamId, Math.max(currentStreamTail, streamTailAtCP));
+
+                // The trim mark is part of the address space information and is also required so clients can observe
+                // updates to streams that have been completely checkpointed.
+                if (streamsAddressSpaceMap.containsKey(streamId) &&
```

Comment:
We do need it, because even if the bitmap is not empty, we need the trim mark for knowing when we can resolve snapshot transactions or not. Suppose you have a S1 with updates at addresses 2, 4, 5, 10... you checkpoint S1 at 10 and trimmed at 4... now your bitmap looks like: 5, 10... if you attempt to do a snapshot transaction at 5, you need to know your trim mark, otherwise you will incorrectly only load 5 only (as the checkpoint is discarded as it has the state up to 10)...